Block Bindings: Add support to image caption attribute in block bindings

SantosGuillamot opened the pull request:
## What?
Add support for the image caption attribute in block bindings.

## Why?
There is an issue caused by this: https://github.com/WordPress/gutenberg/issues/62287.

## How?

These are the main changes made to the pull request:

**Change the core bindings logic**
The [core pull request](https://github.com/WordPress/wordpress-develop/pull/6838) adds the logic to support caption in the bindings replacement logic. Right now, it uses an anonymous class simulating `set_inner_text` until the HTML API provides a similar method. It is done this way to avoid using regex.

**Change the image render logic**
This pull request changes the image render logic to cover two use cases:
* Remove the figcaption element when it exists, and the binding value is empty.
* Add the figcaption element when it doesn't exist, and the binding value is not empty.

This is done again creating an anonymous class with a couple of methods until this can be easily done with HTML API core code. It is done this way to avoid using regex.

**Change the compat folder for versions previous to 6.5**
Apart from that, this pull request adapts the code that runs when a WordPress version below 6.5 is used. They look like big changes, but most of them are just code reorganization. It covers:
* Copy the replacement code used in core.
* Reorganize the code to abstract `gutenberg_is_valid_block_for_block_bindings` and `gutenberg_is_valid_block_binding`.
* Create a new `render_block_data` filter to modify the attributes with the binding value. This is [done in core from 6.5](https://github.com/WordPress/wordpress-develop/blob/c00ad5aa48585311aa6809050c7b079cfe1b9f95/src/wp-includes/class-wp-block.php#L465-L470), but there wasn't an equivalent in Gutenberg compat, and it is needed for this use case.

## Testing Instructions

Keep in mind that in order to test it you need the code from [this core pull request](https://github.com/WordPress/wordpress-develop/pull/6838) apart from this one.

Mainly, three scenarios need to be tested in three different environments:

**Scenarios**
1. An image with existing figcaption is replaced with the pattern overrides value.
2. An image without figcaption creates a figcaption when provided by pattern overrides.
3. An image with existing figcaption is removed when the pattern overrides value is empty.

**Enviroments**
* Test with the branch from this [core pull request](https://github.com/WordPress/wordpress-develop/pull/6838).
* Test with `6.5` branch in core.
* Test with `6.4` branch in core.

**How to test it**

Best way to test it, after having a local environment with the relevant Gutenberg and core branches:
1. Go to a page and add a row with three images: One with a figcaption that will be replaced, another one without figcaption, and other with a caption that will be removed. You can use this snippet:

```html
<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"><!-- wp:image {"width":"126px","height":"auto","sizeSlug":"large","metadata":{"bindings":{"__default":{"source":"core/pattern-overrides"}},"name":"Bulbasur Override"}} -->
<figure class="wp-block-image size-large is-resized"><img src="https://img.pokemondb.net/artwork/large/bulbasaur.jpg" alt="" style="width:126px;height:auto"/><figcaption class="wp-element-caption">Replace</figcaption></figure>
<!-- /wp:image -->

<!-- wp:image {"width":"107px","height":"auto","sizeSlug":"large","metadata":{"bindings":{"__default":{"source":"core/pattern-overrides"}},"name":"Charmander Override"}} -->
<figure class="wp-block-image size-large is-resized"><img src="https://img.pokemondb.net/artwork/large/charmander.jpg" alt="" style="width:107px;height:auto"/></figure>
<!-- /wp:image -->

<!-- wp:image {"width":"118px","height":"auto","sizeSlug":"large","metadata":{"bindings":{"__default":{"source":"core/pattern-overrides"}},"name":"Squirtle Override"}} -->
<figure class="wp-block-image size-large is-resized"><img src="https://img.pokemondb.net/artwork/large/squirtle.jpg" alt="" style="width:118px;height:auto"/><figcaption class="wp-element-caption">Remove</figcaption></figure>
<!-- /wp:image --></div>
<!-- /wp:group -->
```

2. Click on the options and select create pattern, and give it a name.
3. Change the name of the first caption to whatever you want. "Bulbasur", for example.
4. Add a caption to the second image. "Charmander", for example.
5. Remove the caption of the third image.
6. Save and check the frontend shows the expected results.

![Screenshot 2024-06-20 at 18 09 50](https://github.com/WordPress/gutenberg/assets/34552881/47f3c6ca-a72e-4ee8-9217-d258bd9ab020)

7. Switch core branches to ensure it works in all the scenarios.

### Issues before this pull request

https://github.com/WordPress/gutenberg/assets/34552881/f2bb13c6-edbe-4bb2-af7a-ad336815f2e1

As can be seen in the video:
* when you have a synced pattern with an image without caption, you add it to a page, and replace it with an image with caption, the caption is not added and there is no way to add it.
* when you have a synced pattern with an image with caption, you add it to a page, and replace it with an image without caption, the original caption can't be removed.

### Issues fixed after this pull request

https://github.com/WordPress/gutenberg/assets/34552881/5eb76325-a3f4-4831-8995-98aac873d4bc

As can be seen in the video:
* when you have a synced pattern with an image without caption, you add it to a page, and replace it with an image with caption, the caption IS directly added.
* when you have a synced pattern with an image with caption, you add it to a page, and replace it with an image without caption, the original caption CAN be removed. It is not removed by default, but that happens for all the images, not just pattern overrides: [link](https://github.com/WordPress/gutenberg/issues/62468).

### Image caption can be replaced, added, or removed from the editor

https://github.com/WordPress/gutenberg/assets/34552881/7ea617bc-a71c-491a-a4fe-3e0230f6e33e

As can be seen in the video, the following scenarios are possible after this PR:
* You can replace the caption of the original image.
* You can add a new caption if the original image doesn't have one.
* You can remove the caption of the original image.

carolinan commented:
if it is not planned can you please remove it from the board?

afercia commented:
See related https://github.com/WordPress/gutenberg/issues/62287 and https://github.com/WordPress/gutenberg/issues/59819

SantosGuillamot commented:
Unfortunately, it didn't make it in time for 6.6. We focused on other aspects related to the block bindings API, and the approach taken in this draft pull request would need to be discussed in detail. Additionally, after postponing it, we might be able to rely more on the HTML API.

I just removed it from the board.

priethor commented:
As Andrea points out, not supporting this results in bugs in Pattern Overrides, so it's fair to catalog this as a bugfix for 6.6 rather than an enhancement.

SantosGuillamot commented:
Okay, I will work on it during the beta phase then and try to support the image attributes to solve those issues. This is what I have in mind so far:
- Logic to support caption without regex.
- Logic to support link attributes without regex.
- Make it work when the caption doesn’t exist, and it is added dynamically with bindings.
- Make it work when the link doesn’t exist, and it is added dynamically with bindings.
- Allow/disallow editing in the UI as we are doing for the rest of the attributes.
- Probably in a separate pull request, adapt the UI for pattern overrides.
- We might need a backward compatibility `render_block_data` filter to manage the attributes properly.

Apart from that, we are working on the compatibility code here, but these changes would also be needed in the core.

SantosGuillamot commented:
I've been exploring the different options to cover the wanted use cases, and I am not sure which one is better. I am not convinced about either of them:

1. Change the `save.js` HTML and always return the `<a>` wrapper and the `<figcaption>`.

Following a similar approach to [this pull request](https://github.com/WordPress/gutenberg/pull/62220/), I tried to always return the link element wrapper and the figcaption and remove them in the render if they should be empty.

You can see the code [here](https://github.com/WordPress/gutenberg/pull/61255/files#diff-1e4aa29ea183efda9c795271dc1acf6e8049355db104d0dca5b0ea5caf7f619cR85-R121).

2. Don't change the `save.js` and rebuild OR remove the HTML in the render.

Another approach could be keeping the save logic as it is and add logic to the render to handle both case:
* Rebuild the HTML to include the figcaption or the link wrapper if the attributes are changed by bindings.
* Remove the figcaption or the link wrapper if the value from bindings is empty.

I was testing it [here](https://github.com/WordPress/gutenberg/pull/61255/commits/87b58001e6c551312915fa06079b8c4294684c9c#diff-1e4aa29ea183efda9c795271dc1acf6e8049355db104d0dca5b0ea5caf7f619cR42-R76), although it would also need the removal part.

Any opinions?

---

Apart from that, I started working on [this branch](https://github.com/SantosGuillamot/wordpress-develop/compare/fix/revert-changes-to-bindings-logic...SantosGuillamot:wordpress-develop:add/support-all-image-attributes-in-bindings) to include these changes in core as well. Right now, the files we are modifying are just part of the compat folder for versions older than 6.5, where bindings in core didn't exist.

talldan commented on lib/compat/wordpress-6.5/blocks.php:
To test the changes in this file I had to ensure the `WP_Block_Bindings_Registry` isn't loaded in core, as this condition is on line 50:
```
if ( ! class_exists( 'WP_Block_Bindings_Registry' ) ) {
```

However, when I do that, the `gutenberg_block_bindings_replace_html` function runs after the block has rendered (whereas in core the equivalent function runs before the block has rendered), so by that point the caption is being removed by the block's render_callback (if it's empty) and can't be processed.

It's a bit of a conundrum, I think to be able to test properly the gutenberg code needs to run earlier, but I can't see a relevant filter to hook into that would make it work. 🤔 

Perhaps the PHP changes can be made in core first, though I'm not sure what that means for gutenberg supporting older versions of WordPress. Alternatively we could add a new filter in core to hook into as part of the change.

talldan replied:
Lets leave this for a follow-up discussion, as I think the impact is wider than this PR. Probably not something that's easily changed right now.

talldan commented:
> I've been exploring the different options to cover the wanted use cases, and I am not sure which one is better. I am not convinced about either of them

I personally prefer the first option, purely because it keeps the block responsible for the exact markup of the caption. With the second option the block bindings code is taking lots of responsibility for the block's html. I feel it's better if the bindings code is only lightly updating the html as much as possible.

BTW, I pushed a couple of commits, the first one updates another `supported_attrs` array that's in the codebase 😬 

The second one allows for editing of captions in the frontend. A similar change will probably also have to be made for the link, though I think it's good to focus on the caption first (and perhaps those changes are split into two separate PRs?).

talldan commented on lib/compat/wordpress-6.5/blocks.php:
```diff
@@ -123,6 +123,16 @@ function gutenberg_block_bindings_replace_html( $block_content, $block_name, str
 						}
+						if ( 'core/image' === $block_name && 'caption' === $attribute_name ) {
+							// TODO: Don't use regex.
+							return preg_replace_callback(
+								'/<figcaption[^>]*>.*?<\/figcaption>/is',
+								function () use ( $amended_content ) {
+									return $amended_content->get_updated_html();
+								},
+								$block_content
+							);
```
I don't think there's a need to use a callback:
```suggestion
							return preg_replace(
								'/<figcaption[^>]*>.*?<\/figcaption>/is',
								$amended_content->get_updated_html(),
								$block_content
							);
```

cbravobernal commented:
> Change the save.js HTML and always return the <a> wrapper and the <figcaption>.

Would this require to create a `v9` in `deprecated.js`? As far as I remember, any markup change inside save.js requires a deprecation, so previous posts blocks won't crash on loading.

cbravobernal commented:
I got the caption working on the editor, but not in the frontend, I'll take a look at that:

![Screenshot 2024-06-13 at 19 01 56](https://github.com/WordPress/gutenberg/assets/37012961/608686ed-8610-43aa-91e4-aa43f7a22cf6)

talldan commented:
> Would this require to create a v9 in deprecated.js? As far as I remember, any markup change inside save.js requires a deprecation, so previous posts blocks won't crash on loading.

That's a good question. It doesn't cause any noticeable issues when testing, but it looks like that's because the saved markup matches the latest deprecation for the image block, so it successfully upgrades. It does seem to work fine, but it might cause unexpected issues, so it would probably be best to add a deprecation.

SantosGuillamot commented:
I've been thinking more about this and made a few changes. These are some comments about them:
* I've cleaned the pull request to only include caption, which seems to be the main issue here.
* I agree that the changes should be made in core first, and just update the compat and the image rendering in this pull request. I've started a draft pull request for that: [link](https://github.com/WordPress/wordpress-develop/pull/6838). Both PRs should be tested together.
* I've added some logic to cover the uses cases for the image render, without modifying the `save.js` file. To me, it seems a bit safer to handle it this way, because we aren't changing the way it works right now.
* I'm using anonymous classes to extend the Tag Processor with some methods until the HTML API provide those by default. This way, regex aren't needed, it should be more reliable (if they are correct), and those methods remain private.
* I've changed the title and description to reflect all these changes.

talldan commented:
This is working well in my testing so far. I noticed when editing an overriden image in a pattern instance that the toolbar button for toggling the caption is hidden (it's hidden in `contentOnly` mode for some reason that I don't understand). I've pushed a commit that makes it visible, so that users can toggle the caption on and off in patterns:
![Screenshot 2024-06-17 at 5 31 55 pm](https://github.com/WordPress/gutenberg/assets/677833/82db00e2-2ba0-4d86-a2b1-d82f051680db)

cbravobernal commented:
Using the remove caption button is killing the default value or the edited one. Will take a look at it.

https://github.com/WordPress/gutenberg/assets/37012961/46679398-e5c8-4a3b-812b-a6221e43ba6b

talldan commented:
> Using the remove caption button is killing the default value or the edited one. Will take a look at it.

I've been thinking on this, and I wonder if a solution is for the caption to be set to a RichText value of `''` instead of `undefined`. I think this is the line of code:
https://github.com/WordPress/gutenberg/blob/33854e1655710ac4ac33871f787a12c1fd2c85cf/packages/block-library/src/utils/caption.js#L84

My thinking that an empty string could still override a default value, and for regular use of the block it should still serialize to the same thing as `undefined` in the HTML.

The negative is that I imagine it'll save a value of `''` to block binding sources like post meta when a user might want to clear a value. This is why pattern overrides have a 'Reset' button for allowing a user to switch back to the default values.

Have you had any other similar issues to this for block bindings?

cbravobernal commented:
> > Using the remove caption button is killing the default value or the edited one. Will take a look at it.
> 
> I've been thinking on this, and I wonder if a solution is for the caption to be set to a RichText value of `''` instead of `undefined`. I think this is the line of code:
> 
> https://github.com/WordPress/gutenberg/blob/33854e1655710ac4ac33871f787a12c1fd2c85cf/packages/block-library/src/utils/caption.js#L84
> 
> My thinking that an empty string could still override a default value, and for regular use of the block it should still serialize to the same thing as `undefined` in the HTML.
> 
> The negative is that I imagine it'll save a value of `''` to block binding sources like post meta when a user might want to clear a value. This is why pattern overrides have a 'Reset' button for allowing a user to switch back to the default values.
> 
> Have you had any other similar issues to this for block bindings?

I think is the first time we deal with it. Nullifying instead of undefined removes the string, maybe that's better.

https://github.com/WordPress/gutenberg/assets/37012961/b9056094-8e48-49f6-8cdf-8a65ddbca09e

talldan commented:
It's still working well for me in testing, really nice work here! It looks like there are a few changes that need to be brought over from https://github.com/WordPress/wordpress-develop/pull/6838, so I can't really approve it just yet.

> I have added a bunch of e2e tests for the caption. They are expected to fail until https://github.com/WordPress/wordpress-develop/pull/6838 is merged, though.

Given they fail in this way, the tests should probably be moved to WordPress core—they're really asserting that the core code is functioning as expected—the gutenberg version of the bindings processor isn't executed at all from what I can see.

I think this is relevant for any tests that check the 'frontend' output, the editor tests still seem relevant for this repo.

The other feedback is that there have been modifications to the bindings processor in the `lib/compat/wordpress-6.5` folder during the 6.6 cycle, but my understanding is these changes should be in the `wordpress-6.6` folder so that they're supported for two WordPress versions. Having said that, the gutenberg bindings processor isn't executed at all for any WordPress versions after and including 6.5, so the back compat probably isn't working as expected. Perhaps it's ok as it is, but it might worth asking the question to a wider group.

I think it's ok the ship the PR first before addressing any of this. These are just concerns for the gutenberg plugin, but WordPress 6.6 having caption support is more important in the short term.

SantosGuillamot commented:
> Given they fail in this way, the tests should probably be moved to WordPress core—they're really asserting that the core code is functioning as expected—the gutenberg version of the bindings processor isn't executed at all from what I can see.

You're right the processing deserves some testing in core, I can work on that. On the other hand, I feel the tests added in this pull request are still valuable because they are testing the editor and the [changes made to the image rendering](https://github.com/WordPress/gutenberg/pull/61255/files#diff-1e4aa29ea183efda9c795271dc1acf6e8049355db104d0dca5b0ea5caf7f619cR144-R161).

To fully test this new functionality we need:
* Changes to the processing made in core.
* Changes to the image rendering made in this pull request.

That, summed to the editor tests, it's why I thought they should live in Gutenberg instead of core.

> The other feedback is that there have been modifications to the bindings processor in the lib/compat/wordpress-6.5 folder during the 6.6 cycle, but my understanding is these changes should be in the wordpress-6.6 folder so that they're supported for two WordPress versions.

Maybe 🤔  I must say I am not fully familiar with how these things are usually handled. Let me share some context.

The processing of bindings happens in core because we don't need to rely on hooks, and we can directly do it in the block rendering, which seemed the best approach. I believe it was discussed to work "core-first", and just add compatibility code for previous versions.

This was included in WordPress 6.5, and the code in `lib/compat/wordpress-6.5` was added to support bindings in previous versions of WordPress, where the bindings processing doesn't exist in core.

Now, for new functionalities like adding support for caption. Imagine we add support for it in 6.6. We have different scenarios:

* **WordPress 6.6 with/without Gutenberg**: It will always run core code, so it will ALWAYS support it.
* **WordPress 6.5 without Gutenberg**: It will always run core code, so it will NEVER support it.
* **WordPress 6.5 with Gutenberg**: With the current code, it wouldn't work. We would need to create another compat layer. Do we want to support new functionalities in this scenario? It would imply opening these filters and processing the bindings twice, once in core, and once in Gutenberg. And it would probably apply for later versions of WordPress where the functionality is already supported.
* **WordPress 6.4 with Gutenberg**: With the code of this pull request, it would work for WordPress 6.4 with Gutenberg but not for WordPress 6.5 with Gutenberg, which doesn't seem right.

Taking that into account, the compat code in this PR doesn't feel right. I guess there are two options:
1. **Don't support new functionalities in Gutenberg and just work "core-first".**
2. A**dd compat filters in `lib/compat/wordpress-6.6` just to support these new functionalities**. This would mean processing bindings twice, in core and in Gutenberg filters. Additionally, these filters will probably need to run for all versions, including 6.6 where caption is supported, unless we find a way to differentiate it.

I'm slightly in favor of option 1, but I don't know if it is even allowed according to the guidelines. I'd love to know more thoughts. cc: @gziolo

cbravobernal commented:
> I'm slightly in favor of option 1, but I don't know if it is even allowed according to the guidelines. I'd love to know more thoughts. cc: @gziolo

Guidelines says that Gutenberg is compatible with last two versions of WP. Right now are 6.4 and 6.5. 
![Screenshot 2024-06-19 at 11 47 22](https://github.com/WordPress/gutenberg/assets/37012961/29c8d867-bedc-4b0f-a57c-a15b05519677)

Once 6.6 is released, it should be compatible with 6.5. So I'm afraid we need to make it compatible with **WordPress 6.5 with Gutenberg** option.

SantosGuillamot commented:
The question is more oriented to whether all new functionalities must always go through Gutenberg, even though they are added to core. The new functionalities are compatible with Gutenberg, nothing will break. It is just that they could be added directly to core and not ensuring it work in Gutenberg with old WordPress versions.

cbravobernal commented:
> Is it not possible to add new functionalities directly to core? For me, it seems it will be a burden to keep adding compatibility to Gutenberg each release if we decide to work “core-first”.

I think the problem here is that the editor part is inside Gutenberg, and that Synced Patterns is quite dependent on block bindings.

For a editor project, working "core-first" will be hard. As people with latest WP released and Gutenberg (this case, 6.5 and Gutenberg 18.5) will have a different experience than people with WP trunk and Gutenberg 18.5.

I think this may grow exponentially and reach a level where we don't know what works where, that's my main concern.

SantosGuillamot commented:
You're right that for Pattern Overrides it could become an issue. Maybe we can just create a compat file for 6.6 and discuss that for future functionalities.

gziolo commented on lib/compat/wordpress-6.5/blocks.php:
```diff
@@ -67,6 +67,70 @@ function gutenberg_block_bindings_replace_html( $block_content, $block_name, str
 		switch ( $block_type->attributes[ $attribute_name ]['source'] ) {
 			case 'html':
 			case 'rich-text':
+				if ( 'core/image' === $block_name && 'caption' === $attribute_name ) {
```
I think you will have to move it to `compat/wordpress-6.6/` because that functionality is also missing in WordPress 6.5. If the code stays only here, then after support for WordPress 6.4 gets dropped the polyfill for WP 6.5 will be missing.

gziolo commented:
> Taking that into account, the compat code in this PR doesn't feel right. I guess there are two options:
>
> 1. Don't support new functionalities in Gutenberg and just work "core-first".
> 2.  Add compat filters in lib/compat/wordpress-6.6 just to support these new functionalities. This would mean processing bindings twice, in core and in Gutenberg filters. Additionally, these filters will probably need to run for all versions, including 6.6 where caption is supported, unless we find a way to differentiate it.
I'm slightly in favor of option 1, but I don't know if it is even allowed according to the guidelines. I'd love to know more thoughts.

The problem is that WordPress 6.6 is still not released, so for the next few weeks everyone will see the UI changes with Gutenberg installed but that won't work correctly on the front end. I'm in favor of doing double processing for WP 6.5. In general, that would be the strategy moving forward as we open new attributes, we will have to process them in Gutenberg somehow. Maybe, we should have two compat implementations, the on for WP 6.4 covering everything, the one for WP 6.5 covering only captions for images.

SantosGuillamot commented:
Okay, I've added a few changes to support caption in WordPress 6.5, although it doesn't work for 6.4. The main issue is that the HTML API is developed in core, not in Gutenberg. That means that there are API changes between 6.4, 6.5, and 6.6. That makes creating the needed compatibility layers complex.

For this specific use case, we can manage it (at least for 6.5), but I am not sure how it will work in the future. Let's use a hypothetical example: Imagine that in 6.7, the HTML API supports all tags with new APIs, and we decide to support new blocks relying on those APIs. Adding compatibility for older versions might become pretty complex.

Having said that, for this particular pull request, I moved the logic to the 6.6 folder and just process the caption there, keeping the old logic for older versions. Right now:
* For 6.6 (once core branch gets merged): Caption is only processed in core.
* For 6.5: Caption is processed through the added filters.
* For 6.4: Caption is not supported (at least yet).

I added [this check](https://github.com/WordPress/gutenberg/pull/61255/files#diff-0ab29163dafdd24faf6a2c46448b16e5389969f8bb3eca8d4a47eeea6aadad23R49) to avoid processing the caption in 6.4. That's because, again, there are new APIs in 6.5 that we are using and weren't available in 6.4, so the current code doesn't work. If we need to support that use case as well, I can keep exploring how to do that.

sirreal commented on lib/compat/wordpress-6.6/blocks.php:
```diff
@@ -44,3 +44,221 @@ function gutenberg_replace_pattern_override_default_binding( $parsed_block ) {
+				 * This check is needed to add compatibility for that.
+				 * Related issue: https://github.com/WordPress/wordpress-develop/pull/6625
+				 */
+				if ( '>' === $this->html[ $after_opener_tag ] ) {
+					++$after_opener_tag;
+				}
```
We need to be very careful that this check is correct.

The tag opener _must_ end in `>`:

```html
<div att="foo">bla
```

But the content inside the tag could start with `>`!

```html
<div att="foo">> this starts with a &gt; character!
```

SantosGuillamot replied:
You're right! I just changed it in the latest commit.

I think it was working on my testing because it seems to be transformed to `&xxx`. But checking the ending seems safer.

talldan commented:
> Okay, I've added a few changes to support caption in WordPress 6.5 and 6.4.

Thanks for addressing this. Support for 6.4 was maybe unnecessary, as by the time 6.6 is shipped the plugin will only support 6.6 and 6.5. I think it's ok to bend the rules a little for that given the timelines.

> If we consider both PRs ready before the next release, would it still be okay to merged them for WordPress 6.6? Or are these changes too big to be included?

I don't think the PR is too big in terms of the code that will end up in core. The gutenberg change is a lot bigger than the code that will be backported to core because of the tests and the back compat code that was added. The core PR is pretty tiny, the image block changes would be added when the packages are published, and that's most of it.

The HTML processing parts of the PR I personally don't know enough about to assess the risks. Thoughts from those that are more familiar would be welcome.

If that's ok, then my feeling is that it can be shipped, tested during the RC phase, and it can always be reverted from the release if we're seeing problems—it doesn't seem like a complicated thing to revert using the `IS_GUTENBERG_PLUGIN` flags.

SantosGuillamot commented:
A quick summary with some videos to help with the review and the potential merge:

The main goal of this pull request is to solve the problems shared in this issue: https://github.com/WordPress/gutenberg/issues/62287

* When overriding an image without caption with an image with caption doesn't add the caption.
* When overriding an image with caption with an image without caption, users can't delete the old caption.
* Apart from that, it lets users replace, add, or remove the caption directly from the pattern usage.

I made a few videos to illustrate it:

### Issues before this pull request

https://github.com/WordPress/gutenberg/assets/34552881/f2bb13c6-edbe-4bb2-af7a-ad336815f2e1

As can be seen in the video:
* when you have a synced pattern with an image without caption, you add it to a page, and replace it with an image with caption, the caption is not added and there is no way to add it.
* when you have a synced pattern with an image with caption, you add it to a page, and replace it with an image without caption, the original caption can't be removed.

### Issues fixed after this pull request

https://github.com/WordPress/gutenberg/assets/34552881/5eb76325-a3f4-4831-8995-98aac873d4bc

As can be seen in the video:
* when you have a synced pattern with an image without caption, you add it to a page, and replace it with an image with caption, the caption IS directly added.
* when you have a synced pattern with an image with caption, you add it to a page, and replace it with an image without caption, the original caption CAN be removed. It is not removed by default, but if that's the desired behavior, it could be addressed in another pull request. I am not sure if that should be the case.

### Image caption can be replaced, added, or removed from the editor

https://github.com/WordPress/gutenberg/assets/34552881/7ea617bc-a71c-491a-a4fe-3e0230f6e33e

As can be seen in the video, the following scenarios are possible after this PR:
* You can replace the caption of the original image.
* You can add a new caption if the original image doesn't have one.
* You can remove the caption of the original image.

gziolo commented:
The PR got updated in a wat that can be safely merged in the Gutenberg plugin with full backward compatibility for older WP versions. In effect, we can land it first after review is finished and continue discussing whether it matches the expectations raised during testing Pattern Overrides and therefore is required to be included in WordPress 6.6.

cbravobernal commented on lib/compat/wordpress-6.6/blocks.php:
```diff
@@ -44,3 +44,237 @@ function gutenberg_replace_pattern_override_default_binding( $parsed_block ) {
+		$block_reader->set_bookmark( 'figure' );
+	}
+
+		$new_value = wp_kses_post( $source_value );
```
Maybe we can move this escaping to `gutenberg_set_content_between_figcaption_balanced_tags` and `gutenberg_append_figcaption_element_after_tag`.
It will be repeated in two functions, but we will follow the [escape late approach](https://developer.wordpress.org/apis/security/escaping/#toc_3).

cbravobernal commented on lib/compat/wordpress-6.6/blocks.php:
```diff
@@ -44,3 +44,237 @@ function gutenberg_replace_pattern_override_default_binding( $parsed_block ) {
+		 * @return bool Whether the element was properly appended.
+		 */
+		public function gutenberg_append_figcaption_element_after_tag( $new_element ) {
+			$tag_name = $this->get_tag();
```
Should we escape `$new_element` here?

We are passing `<figcaption html>$attributes['caption']</figcaption>`.

cbravobernal commented on packages/block-library/src/image/index.php:
```diff
@@ -22,8 +22,115 @@ function render_block_core_image( $attributes, $content, $block ) {
+		 * @return bool Whether the element was properly appended.
+		 */
+		public function append_figcaption_element_after_tag( $new_element ) {
+			$tag_name = $this->get_tag();
```
Same here https://github.com/WordPress/gutenberg/pull/61255/files#r1648912442

ellatrix commented:
> I'm quite worried about introducing caption support this late in the release cycle. Imo it deserves some testing as a feature in the GB plugin first. I've made this alternative PR that disallows overrides for an image with caption: https://github.com/WordPress/gutenberg/pull/62747.

https://github.com/WordPress/wordpress-develop/pull/6838#issuecomment-2182784356

cbravobernal commented:
Are we ready to merge this one @SantosGuillamot ?

gziolo commented on backport-changelog/6.6/6838.md:
```diff
@@ -0,0 +1,3 @@
+https://github.com/WordPress/wordpress-develop/pull/6838
```
This should now target WP 6.7.

SantosGuillamot commented:
I wanted to make some changes before merging now that we are targeting 6.7. Those mainly being:
* Address the remaining feedback on this pull request.
* Address the remaining feedback on the pull request in core and backport the changes.
* Change the compat files to 6.7.

SantosGuillamot committed:
Transform figure > a selector into a

SantosGuillamot committed:
Use regex (temporary)

SantosGuillamot committed:
Adapt Gutenberg compatibility filters

SantosGuillamot committed:
Add support in the PHP array

SantosGuillamot committed:
Modify the image rendering if bindings are set

SantosGuillamot committed:
Fix regex

SantosGuillamot committed:
Add support in the JS hook

SantosGuillamot committed:
Add support for image caption

SantosGuillamot committed:
Another approach: Remove tags in render

talldan committed:
Update supported binding attribtues in another file

SantosGuillamot committed:
Remove comment

SantosGuillamot committed:
Support only caption

SantosGuillamot committed:
Add `set_inner_text` private method instead of using regex

SantosGuillamot committed:
Fix typo

SantosGuillamot committed:
Only run `set_inner_text` when figcaption exists

talldan committed:
Update client side pattern override supported attributes

SantosGuillamot committed:
Go back to existing `save.js`

cbravobernal committed:
Hide caption controls if the caption is binded

cbravobernal committed:
Remove user permissions check

SantosGuillamot committed:
Fix bookmarks

SantosGuillamot committed:
Fix compatibility with older versions

SantosGuillamot committed:
Add caption e2e tests

SantosGuillamot committed:
Remove extra return

SantosGuillamot committed:
Copy latest changes from core

SantosGuillamot committed:
Add support for 6.4

SantosGuillamot committed:
Only process image caption in 6.6 compat

SantosGuillamot committed:
Remove release_bookmark

SantosGuillamot committed:
Remove extra spaces

SantosGuillamot committed:
Add backport changelog

SantosGuillamot committed:
Change conditional

SantosGuillamot committed:
Clarify methods only apply to figcaption

SantosGuillamot committed:
Move logic to 6.6 compat folder

SantosGuillamot committed:
Make filter caption specific

SantosGuillamot committed:
Don't grab bookmarks until the end

SantosGuillamot committed:
Add warning

SantosGuillamot committed:
Move changelog to 6.7 folder

SantosGuillamot committed:
Remove check to always run caption processing

SantosGuillamot committed:
Only process caption in WP 6.5

SantosGuillamot committed:
Remove caption restriction in pattern overrides

ellatrix commented:
What is the long term plan here? Are we going to keep making exceptions for some blocks and some attributes? Imo, a more long term solution should be explored, for example templates in the block API, so the server knows how to serialise the block.

SantosGuillamot commented:
> What is the long term plan here? Are we going to keep making exceptions for some blocks and some attributes?

The plan is to use the HTML API once it is ready to create an abstraction that works for all block attributes. The main functionalities missing are:
* Support for CSS selectors.
* A `set_inner_html` function or something similar.
* A `set_outer_html` function or something similar.

Once that is in place, we should be able to create an abstraction that works for all use cases just by reading the type and selector of each block attribute.

ellatrix commented:
@SantosGuillamot how does that solve adding or removing attributes?

SantosGuillamot commented:
> how does that solve adding or removing attributes?

Oh, sorry. I thought you meant exceptions in the logic used to replace the HTML based on the block attributes: [link](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R336). I assume you are talking about the changes we are making to the image render (or other blocks), right?

I totally agree that we should explore a long-term solution for that. To be honest, I don't have a clear idea how it could be solved, so any ideas are more than welcome. The initial plan was to fully explore potential paths before opening block bindings for more blocks.

>  templates in the block API

Could you elaborate more on what you mean by templates in the block API? I'd love to understand that better.

ellatrix commented:
> Oh, sorry. I thought you meant exceptions in the logic used to replace the HTML based on the block attributes: [link](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R336). I assume you are talking about the changes we are making to the image render (or other blocks), right?

I'm talking about everything :)

> Could you elaborate more on what you mean by templates in the block API? I'd love to understand that better.

Some sort of templating language that works both client and server side. Maybe something like what @youknowriad [created](https://github.com/youknowriad/blocky?tab=readme-ov-file#usage) a while ago. But we probably need logic in templates. Maybe something like Liquid?

SantosGuillamot commented:
> Some sort of templating language that works both client and server side.

That would definitely help with this particular problem, and I'd love to see something like that in the future. If I am not mistaken, a templating language was also discussed at some point as a possibility while working on the Interactivity API. Maybe @luisherranz has some insights.

cbravobernal commented:
> > Some sort of templating language that works both client and server side.
> 
> That would definitely help with this particular problem, and I'd love to see something like that in the future. If I am not mistaken, a templating language was also discussed at some point as a possibility while working on the Interactivity API. Maybe @luisherranz has some insights.
> 
> EDIT: Thinking about it a bit more, it could have other use cases and it seems a big topic. Maybe it deserves a separate issue/discussion to start gathering feedback?

It has indeed an issue from 5 years ago:
https://github.com/WordPress/gutenberg/issues/14446

luisherranz commented:
> Some sort of templating language that works both client and server side

Yes, we want to start experimenting with that 🙂

gziolo commented:
> > What is the long term plan here? Are we going to keep making exceptions for some blocks and some attributes? Imo, a more long term solution should be explored, for example templates in the block API, so the server knows how to serialise the block.

> EDIT: Thinking about it a bit more, it could have other use cases and it seems a big topic. Maybe it deserves a separate issue/discussion to start gathering feedback?

Indeed, that deserves a better place to discuss all the details, as using any templating solution might not be straightforward with the currently existing extensibility model. There is a lot of nuances even in how the block editor injects additional class names and styles to the serialized HTML with block supports. If we were to use a templating system, we would have to ensure that hooks offered on the client, have a matching implementation on the server so the reconstructed HTML from attributes and the template wouldn't miss any details. 

I also wanted to highlight the proposal from @dmsnell about HTML API templating:  https://make.wordpress.org/core/2023/08/19/progress-report-html-api/#templating. Prototype: https://github.com/WordPress/wordpress-develop/pull/5949. That's something that could be re-implemented in JavaScript. However, it's also a good reminder that conditions and loops are very important aspects here. Overall, it's a bigger topic that this PR only surfaces by trying to address the limitations on the current saving mechanism for static blocks.
